jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CallbackCompletionHandler.java

```diff
@@ -0,0 +1,875 @@
+    }
+
+    /**
+     * @return whether the {@link #handle(Request, Response, Callback)} callback is completed
+     * in case the {@link #getHandlerCallbackTimeout()} expires
+     * @see #getHandlerCallbackTimeout()
+     */
+    @ManagedAttribute("Whether the handle() callback is completed in case of timeout")
+    public boolean isCompleteHandlerCallbackAtTimeout()
+    {
+        return completeHandlerCallbackAtTimeout;
+    }
+
+    public void setCompleteHandlerCallbackAtTimeout(boolean completeHandlerCallbackAtTimeout)
+    {
+        this.completeHandlerCallbackAtTimeout = completeHandlerCallbackAtTimeout;
+    }
+
+    /**
+     * @return the timeout for the execution of the demand callback passed to {@link Request#demand(Runnable)}
+     */
+    @ManagedAttribute("The timeout for the execution of the demand callback")
+    public Duration getDemandCallbackTimeout()
+    {
+        return demandCallbackTimeout;
+    }
+
+    public void setDemandCallbackTimeout(Duration timeout)
+    {
+        this.demandCallbackTimeout = Objects.requireNonNull(timeout);
+    }
+
+    /**
+     * @return the timeout for the execution of a {@link Response#write(boolean, ByteBuffer, Callback)} call
+     */
+    @ManagedAttribute("The timeout for the execution of a response write")
+    public Duration getWriteTimeout()
+    {
+        return writeTimeout;
+    }
+
+    public void setWriteTimeout(Duration timeout)
+    {
+        this.writeTimeout = Objects.requireNonNull(timeout);
+    }
+
+    /**
+     * @return the timeout for the execution of the response write callback passed to {@link Response#write(boolean, ByteBuffer, Callback)}
+     */
+    @ManagedAttribute("The timeout for the execution of the response write callback")
+    public Duration getWriteCallbackTimeout()
+    {
+        return writeCallbackTimeout;
+    }
+
+    public void setWriteCallbackTimeout(Duration timeout)
+    {
+        this.writeCallbackTimeout = Objects.requireNonNull(timeout);
+    }
+
+    @Override
+    public boolean handle(Request originalRequest, Response originalResponse, Callback originalCallback) throws Exception
+    {
+        StateInfo stateInfo = new StateInfo(originalRequest);
+        stateInfos.put(originalRequest, stateInfo);
+        Request.addCompletionListener(originalRequest, x -> stateInfos.remove(originalRequest));
+
+        Request request = originalRequest;
+        if (demandCallbackTimeout != null)
+            request = new RequestWrapper(stateInfo);
+
+        Response response = originalResponse;
+        if (writeTimeout != null || writeCallbackTimeout != null)
+            response = new ResponseWrapper(stateInfo, originalResponse);
+
+        HandlerCallback callback = new HandlerCallback(stateInfo, originalCallback);
+        stateInfo.handlerCallback = callback;
+        // TODO: what if super.handle() throws?
+        boolean handled = super.handle(request, response, callback);
```

Comment:
If `super.handle` throws, we have no idea if the handler is going to call the callback or not... but we do know that whoever catches the throw is very likely to complete the callback.

probably simplest to just remove the StateInfo from the map.... anything else we try to do will be racy.